include/zephyr/sys/cbprintf_cxx.h

```diff
@@ -249,10 +249,12 @@ struct z_cbprintf_cxx_remove_reference < T & > {
 	typedef T type;
 };
 
+#ifndef CONFIG_STD_CPP98
```

Comment:
I would prefer to generalise this using the `__cplusplus` macro since it is more flexible and more or less what we have been using for checking the effective C++ standard in other places.

`__cplusplus >= 201103L`

Reply:
It's not uniformly supported in that way though, xcc defines it as just "1".  Which would actually work in this case (1 is less than 199711), but I doubt that's the only edge case.  What if a compiler supports C++17 but was passed -std=c++11, what do we get?  Do we know for sure?  Have we tested?

Given exactly the kind of nonsense we're dealing with here, I feel strongly that relying on our own abstraction is the right choice.

Reply:
> What if a compiler supports C++17 but was passed -std=c++11, what do we get?

`__cplusplus` should be defined as the effective standard version; so, that would be `201103L` if `-std=c++11`.

> Do we know for sure? Have we tested?

Yes, that is what the standard says and how it must be.

From ISO/IEC 14882:1998(E) 16.8 Predefined macro names:

> The name `__cplusplus` is defined to the value 199711L when compiling a C++
translation unit.

> It's not uniformly supported in that way though, xcc defines it as just "1". Which would actually work in this case (1 is less than 199711)

That is a compiler bug and, as you say, it works either way. You can expect any non-ancient compiler that supports C++11 and above to properly define this macro according to the ISO C++ standard.

Reply:
Changed, though this particular one gives me pretty bad ick.  Relying on compiler conformance for standards minutiae in areas that are **known** to be improperly implemented seems like the opposite of defensive/conservative best practice.